prime-router/src/main/kotlin/ReportStreamFilter.kt

```diff
@@ -20,21 +61,183 @@ fun ReportStreamConditionFilter.codes(): List<String> = this.flatMap { it.codes(
     JsonSubTypes.Type(FHIRExpressionConditionFilter::class, name = "fhirExpression"),
 )
 abstract class ConditionFilter(val value: String) {
-    // TODO: change to .filter() or .passes() method and combine with legacy filter
-    abstract fun codes(): List<String>
+    abstract fun evaluate(
+        bundle: Bundle,
+        filter: Boolean = true,
+    ): ConditionFilterResult
+}
+
+abstract class ObservationConditionFilter(value: String) : ConditionFilter(value) {
+    abstract fun evaluateObservation(bundle: Bundle, observation: Observation): ConditionFilterResult
+
+    override fun evaluate(bundle: Bundle, filter: Boolean): ConditionFilterResult {
+        val passedObservations = mutableListOf<Observation>()
+        var result = bundle.getObservations().fold(ConditionFilterResult()) { result, observation ->
+            val obsResult = evaluateObservation(bundle, observation)
+            if (obsResult.pass) {
+                passedObservations.add(observation)
+            } else if (filter) {
+                bundle.deleteResource(observation)
+            }
+            result.or(obsResult)
+        }
+        if (result.pass && passedObservations.all { it.getMappedConditions().all { code -> code == "AOE" } }) {
+            // TODO: all AOE add appropriate message
```

Comment:
TODO here

Reply:
I wasn't planning on adding any logging here (there isn't in current master) but wanted to double-check.
@arnejduranovic do you think we should log if bundle doesn't get sent to a receiver due to all AOE observations?

We don't currently log anything of the sort when deciding what receivers to route to. Only when pruning.